src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,247 @@
+        Pinning::Pin CreatePin(Execution::Context& context, std::string_view packageId, std::string_view sourceId, const std::string& installedVersion)
+        {
+            if (context.Args.Contains(Execution::Args::Type::GatedVersion))
+            {
+                return Pinning::Pin::CreateGatingPin({ packageId, sourceId }, context.Args.GetArg(Execution::Args::Type::GatedVersion));
+            }
+            else if (context.Args.Contains(Execution::Args::Type::BlockingPin))
+            {
+                return Pinning::Pin::CreateBlockingPin({ packageId, sourceId }, { installedVersion });
+            }
+            else
+            {
+                return Pinning::Pin::CreatePinningPin({ packageId, sourceId }, { installedVersion });
+            }
+        }
+    }
+
+    void OpenPinningIndex(Execution::Context& context)
+    {
+        auto indexPath = Runtime::GetPathTo(Runtime::PathName::LocalState) / "pinning.db";
+
+        AICLI_LOG(CLI, Info, << "Opening pinning index");
+
+        auto pinningIndex = std::filesystem::exists(indexPath) ?
+            PinningIndex::Open(indexPath.u8string(), SQLiteStorageBase::OpenDisposition::ReadWrite) :
+            PinningIndex::CreateNew(indexPath.u8string());
+
+        context.Add<Execution::Data::PinningIndex>(std::make_shared<PinningIndex>(std::move(pinningIndex)));
+    }
+
+    void GetAllPins(Execution::Context& context)
+    {
+        AICLI_LOG(CLI, Info, << "Getting all existing pins");
+        context.Add<Execution::Data::Pins>(context.Get<Execution::Data::PinningIndex>()->GetAllPins());
+    }
+
+    void SearchPin(Execution::Context& context)
+    {
+        auto package = context.Get<Execution::Data::Package>();
+        std::vector<Pinning::Pin> pins;
+
+        // TODO: We should support querying the multiple sources for a package, instead of just one
+        auto availableVersion = package->GetLatestAvailableVersion();
+
+        auto pinningIndex = context.Get<Execution::Data::PinningIndex>();
+        auto pin = pinningIndex->GetPin({
+            availableVersion->GetProperty(PackageVersionProperty::Id).get(),
+            availableVersion->GetProperty(PackageVersionProperty::SourceIdentifier).get() });
+        if (pin)
+        {
+            pins.emplace_back(std::move(pin.value()));
+        }
+
+        context.Add<Execution::Data::Pins>(std::move(pins));
+    }
+
+    // Adds a pin for the current package.
+    // A separate pin will be added for each source.
+    // Required Args: None
+    // Inputs: PinningIndex, Package
+    // Outputs: None
+    void AddPin(Execution::Context& context)
+    {
+        auto package = context.Get<Execution::Data::Package>();
+        auto installedVersion = context.Get<Execution::Data::InstalledPackageVersion>();
+        if (!installedVersion)
+        {
+            AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_NO_APPLICATIONS_FOUND);

```

Comment:
Is there an error code for no "installed" package found. We may want to differentiate that user is trying to pin something not installed vs. package is not found at all.

Or maybe just limit the search behavior to SearchInstalled only so we don't even need this check.

Reply:
If the search behavior is limited, then a user wouldn't be able to add a pin without having the package installed.  The only use case I can think of for pinning a package that isn't installed would be a gating pin so that `winget install` doesn't exceed the gate. But, this can be worked around using `-v` and then adding the pin after install. 

Limiting to search only installed would be good

Reply:
I didn't remember that it is already filtered to only installed as the default search behavior for composite source. Removed this check.